hipchat/hipchat.go

```diff
@@ -275,6 +325,58 @@ func (c *Client) Do(req *http.Request, v interface{}) (*http.Response, error) {
 	return resp, err
 }
 
+func (c *Client) doWithRetryPolicy(req *http.Request, policy RetryPolicy) (*http.Response, error) {
+	currentTry := 0
+
+	for willContinue(currentTry, policy) {
+		currentTry = currentTry + 1
+		resp, err := c.client.Do(req)
+		if err != nil {
+			return nil, err
+		}
+		c.captureRateLimits(resp)
+		if code := resp.StatusCode; code == http.StatusTooManyRequests {
+			if willContinue(currentTry, policy) {
+				doDelay(currentTry, policy)
+			}
+		} else {
+			return resp, nil
+		}
+	}
+	return nil, fmt.Errorf("max retries exceeded (%d)", policy.MaxRetries)
+}
+
+func willContinue(currentTry int, policy RetryPolicy) bool {
+	return currentTry <= policy.MaxRetries
+}
+
+func doDelay(currentTry int, policy RetryPolicy) {
```

Comment:
`doDelay()` name creates ambiguity with request scoped functions like `Do` or `doWithRetryPolicy`. That's maybe dumb but when I first read this I thought the function triggered the request. Maybe you should rename it to `sleep()` ?